### ISSUE TITLE
artifacts

### ISSUE THREAD

**Issue by petaflot** [vispy/vispy#2398]
Drawing some complex geometries with "open" shapes, I notice a large number of artifacts ; this is worsened with 'smooth' shading and low 'fov' values (at zero it's really bad).

In the present case, it makes the solid appear as it if has slit faces, with a strong blinking effect.

https://user-images.githubusercontent.com/32837871/193748748-d4a02a31-c790-40d5-96d0-371bf510f0c1.mp4

on branch main, 04bf1c0a70486650ba8328eabf610802cb4fb554

```
01:00.0 VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Lexa XT [Radeon PRO WX 3200] (rev 10)

DRM_RADEON [=m]
DRM_RADEON_USERPTR [=n]
FB_RADEON [=n]
FB_RADEON_BACKLIGHT [=n]                                                                                                             │   
FB_RADEON_DEBUG [=n]
FB_RADEON_I2C [=n]

Gentoo Base System release 2.8
x11-drivers/xf86-video-amdgpu-22.0.0 (23:18:16 22. 08. 22)
```

**Comment by djhoese** [comment]
@rougier any advice? @petaflot Could you please provide a minimal set of vertices and code that reproduce this issue so it can be played with?

**Comment by rougier** [comment]
The mesh seems to be wrong and I've difficulties interpreting its "real shape". Some artifacts seems to be related to coincidental surfaces (eg. coplanar triangles) while some other are related to bad normals. Depending on the origin of the mesh, I think it needs to be fixed. Can you try to visualize the mesh with another software and check the result?

**Comment by petaflot** [comment]
@djhoese will do. I noticed some other unwanted rendering effects, I will document them as well. just a little busy at the moment..

@rougier : 
- the shape is non-manifold, you may see it as some sort of origami : it's a surface that has two sides and is folded in a way to make it very useful. last time I checked, all the normals (faces and vertices) were pointing the correct way
- believe there are no coincidental surfaces (there may be some triangles that intersect, though) : my software is meant to compute trajectories for a tool head, and the way it's generated it really shouldn't happen. 
- do you have another software to recommend for linux?

**Comment by rougier** [comment]
Do you have a simpler version of the mesh of some way to simplify it ? Or any reference image ?

If you have the mesh file, you can try blender to render it.

**Comment by petaflot** [comment]
and no, I don't have a meshfile... I've never worked with meshed before, so I just wrote (it's not quite finished.. working on interpolation between layers) a function so I can do

```python
         self.points = self.P.get_points()   # ALL the points, required for the mesh
         faces = self.P.get_mesh()

         meshdata = MeshData.MeshData(
                 vertices=self.points[0],
                 faces=faces[0],
                 face_colors = faces[1],
             )
```

I suppose I can then do `file.write( meshdata.save() )` ?

**Comment by petaflot** [comment]
do you have a suggestion for a specification of a suitable format? or an example... I tried to open orig/triceratops.obj.gz from the example but only got a 404

in my script I had computed the normals before I found out vispy does it too, so I don't need vispy for that

**Comment by petaflot** [comment]
found it :-) blender gave me the "Wavefront" clue I needed

**Comment by petaflot** [comment]
I can now export to OBJ (except for face alpha).

with alpha = 1

https://user-images.githubusercontent.com/32837871/196165659-75a7a181-bfe3-4e85-bf2f-bf56bc1f841c.mp4

with alpha = 0 and alpha = 0.8

https://user-images.githubusercontent.com/32837871/196166116-5ede4bf2-e72b-45e2-9186-87323bf01e7a.mp4

and the mesh file

[mesh.zip](https://github.com/vispy/vispy/files/9799978/mesh.zip)

I did not find a way to add the alpha to the mesh file, can you help?

**Comment by petaflot** [comment]
above OBJ shows problem related to transparency. this one shows artifacts on some faces (OK with blender), alpha is 1.

![Untitled](https://user-images.githubusercontent.com/32837871/196175280-368e47bb-d946-4a3d-a52d-c3255eab9b69.png)

(don't bother trying to make sense of the geometry... it's just a work in progress and shouldn't be what it is)

[prismoid_artifact.zip](https://github.com/vispy/vispy/files/9800337/prismoid_artifact.zip)

**Comment by petaflot** [comment]
more artifacts:

![Untitled](https://user-images.githubusercontent.com/32837871/196477133-fa5744ba-70a1-4f70-9ed4-e363fde2e8d8.png)

[truncated_pyramid.zip (OBJ)](https://github.com/vispy/vispy/files/9812198/truncated_pyramid.zip)

same if the normals are inverted, same if the pyramid as 4 sides

**Comment by djhoese** [comment]
@rougier do these images and videos give you any ideas what might be going wrong in vispy? Note the videos don't show up in firefox but chrome seems fine (or if you download them and view them locally).

**Comment by rougier** [comment]
Thanks for the movie. The mesh seems simple enough and vispy should be able to render it. The last image comes from vispy ?

As for transparency, you'll need to use order independent transparency if you want "true" transparency. The reason is that with normal rendering, triangles are drawn one after the other and this introduces some discrepancies when you have transparent triangles.

**Comment by brisvag** [comment]
I think this is familiar... @petaflot, what happens if you simply multiply all the coordinates of the points by, say, 100 (making everything 100 times bigger)?

**Comment by petaflot** [comment]
@rougier last last 2 images and 2 videos are from vispy.

@rougier a clue to user "order independant transparency" ? grepping for code for "transparency" returned almost nothing

@brisvag please bear with me...  I started some heavy changes on my lib (learning to use quaternions!) and at the moment it's unusable with a commit half on the way. should be there in a week or so, maybe 2

**Comment by rougier** [comment]
@petaflot See https://github.com/glumpy/glumpy/blob/master/examples/transparency/commented-oit.py

**Comment by brisvag** [comment]
> As for order independent transparency: it's not something vispy can do yet. There are plans for it, but no real implementation.

> @petaflot See [https://github.com/glumpy/glumpy/blob/master/examples/transparency/commented-oit.py](https://github.com/glumpy/glumpy/blob/master/examples/transparency/commented-oit.py?rgh-link-date=2022-11-28T07%3A41%3A08Z)

Right, let me rephrease that: you can definitely do it with vispy already following a similar process as @rougier's gumpy implementation. However, you can't really do it out of the box with the high level `Visual` interface.

**Comment by petaflot** [comment]
@brisvag thanks for rephrasing, but I have little to no idea what you mean.

Unfortunately I don't have the leisure to really test this now :-( around Christmas-time should do it.

I did execute that code though, and while it seems perfectly fine with a static image, I have this strong artifact more or less at the middle-high position (the position of the artifact stays relatively constant each time the app is started, but can fluctuate). Interestingly enough, **there was no artifact** when doing a screencast with ffmpeg. So I filmed it with my phone (slo-mo 4x, 120FPS, edited with avidemux, upside-down but preferred not to rotate to avoid loses due to re-encoding):


https://user-images.githubusercontent.com/32837871/204263123-2f1c7c0d-83b1-4b6c-a064-edcbd7ca3268.mp4

I tried changing monitor settings (Samsung FreeSync), this changed nothing. Since the screencast with ffmpeg is clean, this is definitely not a vispy issue ; maybe you have suggestions on where I should look/ask, what I could/should try to find the source of the problem and hopefully solve it ?

**Comment by brisvag** [comment]
@petaflot did you try my first suggestion?

> what happens if you simply multiply all the coordinates of the points by, say, 100 (making everything 100 times bigger)?

**Comment by petaflot** [comment]
not yet :-/ I'm in the middle of fixing a few issues and at the same time implementing quaternions, didn't want to commit at this stage, really busy with some other crap (government-related paperwork, really is a pin in the back)

**Comment by rougier** [comment]
It looks like a double-buffering issue (as if you had no double-buffering)
